File: app/src/main/java/com/github/overpass/gather/ui/meeting/join/JoinFragment.java
package com.github.overpass.gather.ui.meeting.join;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;

import com.github.overpass.gather.App;
import com.github.overpass.gather.R;
import com.github.overpass.gather.commons.date.FormatKt;
import com.github.overpass.gather.ui.base.BaseFragmentKt;
import com.github.overpass.gather.ui.create.MeetingType;
import com.github.overpass.gather.ui.dialog.progress.indeterminate.ProgressDialogFragment;
import com.github.overpass.gather.ui.meeting.base.LoadMeetingStatus;
import com.github.overpass.gather.ui.meeting.chat.ChatFragment;
import com.github.overpass.gather.ui.meeting.enrolled.EnrolledActivity;
import com.google.android.material.snackbar.Snackbar;

import org.jetbrains.annotations.NotNull;

import java.text.SimpleDateFormat;
import java.util.Locale;

import butterknife.BindView;
import butterknife.OnClick;

import static com.github.overpass.gather.commons.android.SnackbarKt.snackbar;
import static com.github.overpass.gather.commons.android.fragment.TransactionKt.transaction;

public class JoinFragment extends BaseFragmentKt<JoinViewModel> {

    @BindView(R.id.flProgress)
    FrameLayout flProgress;
    @BindView(R.id.tvName)
    TextView tvName;
    @BindView(R.id.tvAddress)
    TextView tvAddress;
    @BindView(R.id.tvRatio)
    TextView tvRatio;
    @BindView(R.id.tvPrivateMeeting)
    TextView tvPrivateMeeting;
    @BindView(R.id.tvDate)
    TextView tvDate;
    @BindView(R.id.ivMeetingType)
    ImageView ivMeetingType;
    @BindView(R.id.toolbarJoin)
    Toolbar toolbarJoin;
    @BindView(R.id.btnJoin)
    TextView btnJoin;

    @NotNull
    @Override
    protected JoinViewModel createViewModel() {
        return getViewModelProvider().get(JoinViewModel.class);
    }

    @Override
    protected int getLayoutRes() {
        return R.layout.fragment_join;
    }

    @Override
    protected void inject() {
        App.Companion.getComponentManager(this)
                .getJoinComponent()
                .inject(this);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        toolbarJoin.setNavigationOnClickListener(navIcon -> requireActivity().finish());
        flProgress.setVisibility(View.VISIBLE);
        getViewModel().loadMeetingCheckEnrolled()
                .observe(getViewLifecycleOwner(), this::handleLoadStatus);
    }

    @OnClick(R.id.btnJoin)
    public void onJoinClick() {
        getViewModel().join().observe(getViewLifecycleOwner(), this::handleJoin);
    }

    private void handleJoin(JoinStatus status) {
        switch (status.tag()) {
            case JoinStatus.ENROLLED:
                handleEnrolled(status.as(JoinStatus.Enrolled.class));
                break;
            case JoinStatus.ERROR:
                handleJoinError(status.as(JoinStatus.Error.class));
                break;
            case JoinStatus.JOINED:
                handleJoined(status.as(JoinStatus.Joined.class));
                break;
            case JoinStatus.PROGRESS:
                handleProgress(status.as(JoinStatus.Progress.class));
                break;
        }
    }

    private void handleProgress(JoinStatus.Progress progress) {
        ProgressDialogFragment.Companion.show(getFragmentManager());
    }

    private void handleJoined(JoinStatus.Joined joined) {
        ProgressDialogFragment.Companion.hide(getFragmentManager());
        transaction(this)
                .replace(R.id.flMeetingContainer, ChatFragment.newInstance())
                .commit();
    }

    private void handleJoinError(JoinStatus.Error error) {
        ProgressDialogFragment.Companion.hide(getFragmentManager());
        snackbar(ivMeetingType, error.getThrowable().getLocalizedMessage(), Snackbar.LENGTH_SHORT);
    }

    private void handleEnrolled(JoinStatus.Enrolled enrolled) {
        ProgressDialogFragment.Companion.hide(getFragmentManager());
        startActivity(new Intent(getContext(), EnrolledActivity.class));
        requireActivity().finish();
    }

    protected void handleLoadStatus(LoadMeetingStatus loadMeetingStatus) {
        switch (loadMeetingStatus.tag()) {
            case LoadPrivateMeetingStatus.ERROR:
                handleLoadError(loadMeetingStatus.as(LoadPrivateMeetingStatus.Error.class));
                break;
            case LoadPrivateMeetingStatus.SUCCESS:
                handleLoadSuccess(loadMeetingStatus.as(LoadPrivateMeetingStatus.Success.class));
                break;
            case LoadPrivateMeetingStatus.PROGRESS:
                handleProgress(loadMeetingStatus.as(LoadPrivateMeetingStatus.Progress.class));
                break;
        }
    }

    protected void handleLoadSuccess(LoadPrivateMeetingStatus.Success success) {
        flProgress.setVisibility(View.GONE);
        tvName.setText(success.getMeetingAndRatio().getMeeting().getName());
        tvPrivateMeeting.setVisibility(success.getMeetingAndRatio().getMeeting().isPrivate()
                ? View.VISIBLE : View.INVISIBLE);
        if (success.isAlreadyEnrolled()) {
            tvPrivateMeeting.setText(R.string.you_have_already_enrolled);
            btnJoin.setVisibility(View.GONE);
        }
        String ratio = String.format(Locale.getDefault(), "%d / %d",
                success.getMeetingAndRatio().getRatio().getCurrent(),
                success.getMeetingAndRatio().getRatio().getMax());
        tvRatio.setText(ratio);
        SimpleDateFormat format = FormatKt.meetingDateFormat();
        tvDate.setText(format.format(success.getMeetingAndRatio().getMeeting().getDate()));
        if (MeetingType.isBusiness(success.getMeetingAndRatio().getMeeting().getType())) {
            ivMeetingType.setImageResource(R.drawable.ic_case_large);
        } else if (MeetingType.isProtest(success.getMeetingAndRatio().getMeeting().getType())) {
            ivMeetingType.setImageResource(R.drawable.ic_bund_large);
        } else {
            ivMeetingType.setImageResource(R.drawable.ic_beer_large);
        }
        getViewModel().getAddress(success.getMeetingAndRatio().getMeeting().getLongitude(),
                success.getMeetingAndRatio().getMeeting().getLatitude())
                .observe(getViewLifecycleOwner(), tvAddress::setText);
    }

    protected void handleProgress(LoadPrivateMeetingStatus.Progress progress) {
        flProgress.setVisibility(View.VISIBLE);
    }

    protected void handleLoadError(LoadPrivateMeetingStatus.Error error) {
        flProgress.setVisibility(View.GONE);
    }

    public static JoinFragment newInstance() {
        return new JoinFragment();
    }
}

File: app/src/main/java/by/overpass/gather/ui/meeting/join/JoinFragment.kt
package by.overpass.gather.ui.meeting.join

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.lifecycle.Observer
import by.overpass.gather.App.Companion.componentManager
import by.overpass.gather.R
import by.overpass.gather.commons.android.fragment.transaction
import by.overpass.gather.commons.android.snackbar
import by.overpass.gather.commons.date.meetingDateFormat
import by.overpass.gather.ui.base.BaseFragmentKt
import by.overpass.gather.ui.create.MeetingType
import by.overpass.gather.ui.dialog.progress.indeterminate.ProgressDialogFragment
import by.overpass.gather.ui.meeting.base.LoadMeetingStatus
import by.overpass.gather.ui.meeting.chat.ChatFragment
import by.overpass.gather.ui.meeting.enrolled.EnrolledActivity
import com.google.android.material.snackbar.Snackbar
import kotlinx.android.synthetic.main.fragment_join.*
import java.util.*

class JoinFragment : BaseFragmentKt<JoinViewModel>() {

    override fun createViewModel(): JoinViewModel {
        return viewModelProvider.get(JoinViewModel::class.java)
    }

    override fun getLayoutRes(): Int {
        return R.layout.fragment_join
    }

    override fun inject() {
        componentManager.getJoinComponent()
                .inject(this)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        toolbarJoin.setNavigationOnClickListener { navIcon -> requireActivity().finish() }
        flProgress.visibility = View.VISIBLE
        viewModel.loadMeetingCheckEnrolled()
                .observe(viewLifecycleOwner, Observer<LoadPrivateMeetingStatus> { this.handleLoadStatus(it) })
        btnJoin.setOnClickListener {
            viewModel.join().observe(viewLifecycleOwner, Observer<JoinStatus> { this.handleJoin(it) })
        }
    }

    private fun handleJoin(status: JoinStatus) {
        when (status.tag()) {
            JoinStatus.ENROLLED -> handleEnrolled(status.`as`<JoinStatus.Enrolled>(JoinStatus.Enrolled::class.java))
            JoinStatus.ERROR -> handleJoinError(status.`as`<JoinStatus.Error>(JoinStatus.Error::class.java))
            JoinStatus.JOINED -> handleJoined(status.`as`<JoinStatus.Joined>(JoinStatus.Joined::class.java))
            JoinStatus.PROGRESS -> handleProgress(status.`as`<JoinStatus.Progress>(JoinStatus.Progress::class.java))
        }
    }

    private fun handleProgress(progress: JoinStatus.Progress) {
        ProgressDialogFragment.show(fragmentManager)
    }

    private fun handleJoined(joined: JoinStatus.Joined) {
        ProgressDialogFragment.hide(fragmentManager)
        transaction()
                .replace(R.id.flMeetingContainer, ChatFragment.newInstance())
                .commit()
    }

    private fun handleJoinError(error: JoinStatus.Error) {
        ProgressDialogFragment.hide(fragmentManager)
        ivMeetingType.snackbar(error.throwable.localizedMessage, Snackbar.LENGTH_SHORT)
    }

    private fun handleEnrolled(enrolled: JoinStatus.Enrolled) {
        ProgressDialogFragment.hide(fragmentManager)
        startActivity(Intent(context, EnrolledActivity::class.java))
        requireActivity().finish()
    }

    protected fun handleLoadStatus(loadMeetingStatus: LoadMeetingStatus) {
        when (loadMeetingStatus.tag()) {
            LoadPrivateMeetingStatus.ERROR -> handleLoadError(loadMeetingStatus.`as`<LoadPrivateMeetingStatus.Error>(LoadPrivateMeetingStatus.Error::class.java))
            LoadPrivateMeetingStatus.SUCCESS -> handleLoadSuccess(loadMeetingStatus.`as`<LoadPrivateMeetingStatus.Success>(LoadPrivateMeetingStatus.Success::class.java))
            LoadPrivateMeetingStatus.PROGRESS -> handlePrivateMeetingProgress()
        }
    }

    protected fun handleLoadSuccess(success: LoadPrivateMeetingStatus.Success) {
        flProgress.visibility = View.GONE
        tvName.text = success.meetingAndRatio.meeting.name
        tvPrivateMeeting.visibility = if (success.meetingAndRatio.meeting.isPrivate)
            View.VISIBLE
        else
            View.INVISIBLE
        if (success.isAlreadyEnrolled) {
            tvPrivateMeeting.setText(R.string.you_have_already_enrolled)
            btnJoin.visibility = View.GONE
        }
        val ratio = String.format(Locale.getDefault(), "%d / %d",
                success.meetingAndRatio.ratio.current,
                success.meetingAndRatio.ratio.max)
        tvRatio.text = ratio
        val format = meetingDateFormat()
        tvDate.text = format.format(success.meetingAndRatio.meeting.date)
        if (MeetingType.isBusiness(success.meetingAndRatio.meeting.type)) {
            ivMeetingType.setImageResource(R.drawable.ic_case_large)
        } else if (MeetingType.isProtest(success.meetingAndRatio.meeting.type)) {
            ivMeetingType.setImageResource(R.drawable.ic_bund_large)
        } else {
            ivMeetingType.setImageResource(R.drawable.ic_beer_large)
        }
        viewModel.getAddress(success.meetingAndRatio.meeting.longitude,
                success.meetingAndRatio.meeting.latitude)
                .observe(viewLifecycleOwner, Observer<String> { tvAddress.text = it })
    }

    protected fun handlePrivateMeetingProgress() {
        flProgress.visibility = View.VISIBLE
    }

    protected fun handleLoadError(error: LoadPrivateMeetingStatus.Error) {
        flProgress.visibility = View.GONE
    }

    companion object {

        fun newInstance(): JoinFragment {
            return JoinFragment()
        }
    }
}
